icepack/models/hybrid.py

```diff
@@ -107,40 +107,54 @@ def terminus(**kwargs):
     mesh = u.ufl_domain()
     zdegree = u.ufl_element().degree()[1]
 
-    ζ = firedrake.SpatialCoordinate(mesh)[2]
+    ζ = firedrake.SpatialCoordinate(mesh)[mesh._geometric_dimension-1]
+
     b = s - h
     ζ_sl = firedrake.max_value(-b, 0) / h
     p_W = ρ_W * g * h * _pressure_approx(zdegree + 1)(ζ, ζ_sl)
     p_I = ρ_I * g * h * (1 - ζ)
 
-    ν = facet_normal_2(mesh)
+    ν = facet_normal_nd(mesh)
+
     return (p_I - p_W) * inner(u, ν) * h
 
 
 def stresses(ε_x, ε_z, A):
     r"""Calculate the membrane and vertical shear stresses for the given
     horizontal and shear strain rates and fluidity"""
-    I = Identity(2)
-    tr = trace(ε_x)
-    ε_e = sqrt((inner(ε_x, ε_x) + inner(ε_z, ε_z) + tr**2) / 2)
-    μ = 0.5 * A**(-1 / n) * ε_e**(1 / n - 1)
-    return 2 * μ * (ε_x + tr * I), 2 * μ * ε_z
+    dim = get_mesh_dimensions(ε_x.ufl_domain())
+    if dim in [2, 2.5]:
+        I = Identity(2)
+        tr = trace(ε_x)
+        ε_e = sqrt((inner(ε_x, ε_x) + inner(ε_z, ε_z) + tr**2) / 2)
+        μ = 0.5 * A**(-1 / n) * ε_e**(1 / n - 1)
+        return 2 * μ * (ε_x + tr * I), 2 * μ * ε_z
+    elif dim in [1, 1.5]:
+        ε_e = sqrt((2*inner(ε_x, ε_x) + inner(ε_z, ε_z)) / 2)
+        μ = 0.5 * A**(-1 / n) * ε_e**(1 / n - 1)
+        return 2 * μ * (ε_x + ε_x), 2 * μ * ε_z
 
 
 def horizontal_strain(u, s, h):
     r"""Calculate the horizontal strain rate with corrections for terrain-
     following coordinates"""
-    ζ = firedrake.SpatialCoordinate(u.ufl_domain())[2]
+    mesh = u.ufl_domain()
+    dim = get_mesh_dimensions(mesh)
+    ζ = firedrake.SpatialCoordinate(mesh)[mesh._geometric_dimension-1]
     b = s - h
-    v = -((1 - ζ) * grad_2(b) + ζ * grad_2(s)) / h
-    du_dζ = u.dx(2)
-    return sym(grad_2(u)) + 0.5 * (outer(du_dζ, v) + outer(v, du_dζ))
+    v = -((1 - ζ) * grad_nd(b) + ζ * grad_nd(s)) / h
+    du_dζ = u.dx(mesh._geometric_dimension-1)
+    if dim in [2, 2.5]:
+        return sym(grad_nd(u)) + 0.5 * (outer(du_dζ, v) + outer(v, du_dζ))
+    elif dim in [1, 1.5]:
+        return grad_nd(u) + 0.5 * (outer(du_dζ, v) + outer(v, du_dζ))
```

Comment:
I think this can be simplified further to just `du_dζ * v`. The outer products are there to get us tensors but in the X or XZ models those terms are just scalars.